File: packages/bench/src/advanced-convergence.ts
/**
 * Numerical convergence: columna/advanced vs scipy / numpy on identical inputs.
 *
 *   pnpm --filter @columna/bench run advanced:convergence
 */
import { mkdirSync, writeFileSync } from 'node:fs'
import * as A from 'columna/advanced'

const OUT = new URL('../results/advanced-convergence-columna.json', import.meta.url)
const g = A.random(20260915)

type Case = {
  id: string
  group: string
  label: string
  python: 'scipy' | 'numpy'
  columna: Record<string, number | number[]>
  inputs: Record<string, unknown>
  notes?: string
}

const cases: Case[] = []
const add = (c: Case) => cases.push(c)

// ---- Distributions ---------------------------------------------------------------------------------
{
  const x = Array.from({ length: 200 }, (_, i) => -3 + (6 * i) / 199)
  const u = Array.from({ length: 200 }, (_, i) => 0.001 + (0.998 * i) / 199)
  const xPos = x.map((v) => Math.abs(v) * 2 + 0.01)
  const xGam = x.map((v) => Math.abs(v) * 3 + 0.01)
  add({
    id: 'dist.normal.cdf',
    group: 'Distributions',
    label: 'normal.cdf (200-point grid)',
    python: 'scipy',
    inputs: { x },
    columna: { values: Array.from(A.normal().map('cdf', x)) },
  })
  add({
    id: 'dist.normal.ppf',
    group: 'Distributions',
    label: 'normal.ppf (200-point grid)',
    python: 'scipy',
    inputs: { u },
    columna: { values: Array.from(A.normal().map('ppf', u)) },
  })
  add({
    id: 'dist.t.cdf',
    group: 'Distributions',
    label: 't(10).cdf',
    python: 'scipy',
    inputs: { x },
    columna: { values: Array.from(A.t(10).map('cdf', x)) },
  })
  add({
    id: 'dist.t.ppf',
    group: 'Distributions',
    label: 't(10).ppf',
    python: 'scipy',
    inputs: { u },
    columna: { values: Array.from(A.t(10).map('ppf', u)) },
  })
  add({
    id: 'dist.chi2.sf',
    group: 'Distributions',
    label: 'chi2(5).sf',
    python: 'scipy',
    inputs: { x: xPos },
    columna: { values: Array.from(A.chi2(5).map('sf', xPos)) },
  })
  add({
    id: 'dist.gamma.cdf',
    group: 'Distributions',
    label: 'gamma(2.5, scale=3).cdf',
    python: 'scipy',
    inputs: { x: xGam },
    columna: { values: Array.from(A.gamma(2.5, 3).map('cdf', xGam)) },
  })
  add({
    id: 'dist.beta.ppf',
    group: 'Distributions',
    label: 'beta(2,5).ppf',
    python: 'scipy',
    inputs: { u },
    columna: { values: Array.from(A.beta(2, 5).map('ppf', u)) },
  })
  add({
    id: 'dist.weibull.ppf',
    group: 'Distributions',
    label: 'weibull(1.8,50).ppf',
    python: 'scipy',
    inputs: { u },
    columna: { values: Array.from(A.weibull(1.8, 50).map('ppf', u)) },
  })
  const q = Array.from({ length: 40 }, (_, i) => 0.5 + (4.5 * i) / 39)
  const p = Array.from({ length: 40 }, (_, i) => 0.5 + (0.49 * i) / 39)
  add({
    id: 'dist.ptukey',
    group: 'Distributions',
    label: 'ptukey(q; k=4, df=20)',
    python: 'scipy',
    inputs: { q },
    columna: { values: q.map((qi) => A.ptukey(qi, 4, 20)) },
  })
  add({
    id: 'dist.qtukey',
    group: 'Distributions',
    label: 'qtukey(p; k=4, df=20)',
    python: 'scipy',
    inputs: { p },
    columna: { values: p.map((pi) => A.qtukey(pi, 4, 20)) },
  })
}

// ---- Basic statistics ------------------------------------------------------------------------------
{
  const x1 = Array.from(g.normal(500, 50, 10))
  const a = Array.from(g.normal(400))
  const b = Array.from(g.normal(400, 0.15, 1.1))
  const t1 = A.ttest1(x1, { mu: 50 })
  const t2 = A.ttest2(a, b)
  const z1 = A.ztest1(x1, { sigma: 10, mu: 50 })
  add({
    id: 'ttest1',
    group: 'Basic statistics',
    label: 'one-sample t (n=500)',
    python: 'scipy',
    inputs: { x: x1, mu: 50 },
    columna: { statistic: t1.statistic, pValue: t1.pValue, estimate: t1.estimate },
  })
  add({
    id: 'ttest2',
    group: 'Basic statistics',
    label: 'Welch two-sample t (n=400+400)',
    python: 'scipy',
    inputs: { a, b },
    columna: { statistic: t2.statistic, pValue: t2.pValue, estimate: t2.estimate },
  })
  add({
    id: 'ztest1',
    group: 'Basic statistics',
    label: 'one-sample z (known σ)',
    python: 'numpy',
    inputs: { x: x1, mu: 50, sigma: 10 },
    columna: { statistic: z1.statistic, pValue: z1.pValue },
  })
  const p1 = A.propTest1(350, 1000, { p0: 0.3 })
  add({
    id: 'propTest1',
    group: 'Basic statistics',
    label: 'binomial exact 350/1000 vs 0.3',
    python: 'scipy',
    inputs: { k: 350, n: 1000, p0: 0.3 },
    columna: { pValue: p1.pValue, estimate: p1.estimate },
  })
  const fisher = A.propTest2(120, 400, 90, 380, { method: 'fisher' })
  add({
    id: 'propTest2.fisher',
    group: 'Basic statistics',
    label: 'Fisher exact 2×2',
    python: 'scipy',
    inputs: { table: [[120, 280], [90, 290]] },
    columna: { pValue: fisher.pValue },
  })
  const xa = Array.from(g.normal(300))
  const xb = xa.map((v) => 0.6 * v + g.normal(1)[0]!)
  const cp = A.corrTest(xa, xb)
  const cs = A.corrTest(xa, xb, { method: 'spearman' })
  add({
    id: 'corrTest.pearson',
    group: 'Basic statistics',
    label: 'Pearson r (n=300)',
    python: 'scipy',
    inputs: { a: xa, b: xb },
    columna: { estimate: cp.estimate, statistic: cp.statistic, pValue: cp.pValue },
  })
  add({
    id: 'corrTest.spearman',
    group: 'Basic statistics',
    label: 'Spearman ρ (n=300)',
    python: 'scipy',
    inputs: { a: xa, b: xb },
    columna: { estimate: cs.estimate, pValue: cs.pValue },
  })
  const xv = Array.from(g.normal(200, 0, 2))
  const vt = A.varTest1(xv, { sigma0: 2 })
  add({
    id: 'varTest1',
    group: 'Basic statistics',
    label: 'χ² variance test',
    python: 'scipy',
    inputs: { x: xv, sigma0: 2 },
    columna: { statistic: vt.statistic, pValue: vt.pValue },
  })
}

// ---- ANOVA / tables --------------------------------------------------------------------------------
{
  const groups: Record<string, number[]> = {}
  for (let i = 0; i < 4; i++) groups[`g${i}`] = Array.from(g.normal(80, i * 0.4))
  const an = A.anova(groups)
  const lv = A.levene(groups)
  const bt = A.bartlett(groups)
  add({
    id: 'anova',
    group: 'ANOVA',
    label: 'one-way ANOVA 4×80',
    python: 'scipy',
    inputs: { groups },
    columna: { statistic: an.statistic, pValue: an.pValue },
  })
  add({
    id: 'levene',
    group: 'ANOVA',
    label: 'Levene',
    python: 'scipy',
    inputs: { groups },
    columna: { statistic: lv.statistic, pValue: lv.pValue },
  })
  add({
    id: 'bartlett',
    group: 'ANOVA',
    label: 'Bartlett',
    python: 'scipy',
    inputs: { groups },
    columna: { statistic: bt.statistic, pValue: bt.pValue },
  })
  const table = Array.from({ length: 5 }, () => Array.from(g.integer(5, 8, 40)))
  const chi = A.chi2test(table)
  add({
    id: 'chi2test',
    group: 'Tables',
    label: 'χ² contingency 5×5',
    python: 'scipy',
    inputs: { table },
    columna: { statistic: chi.statistic, pValue: chi.pValue },
  })
  const observed = Array.from(g.integer(10, 20, 60))
  const gof = A.chi2gof(observed)
  add({
    id: 'chi2gof',
    group: 'Tables',
    label: 'χ² GOF (equal expected)',
    python: 'scipy',
    inputs: { observed },
    columna: { statistic: gof.statistic, pValue: gof.pValue },
  })
}

// ---- Normality / nonparametric ---------------------------------------------------------------------
{
  const xn = Array.from(g.normal(500))
  const ad = A.andersonDarling(xn)
  add({
    id: 'andersonDarling',
    group: 'Normality',
    label: 'Anderson–Darling A² (n=500)',
    python: 'scipy',
    inputs: { x: xn },
    columna: { statistic: ad.statistic },
    notes: 'Compare A²; p-value tables may differ from scipy',
  })
  const xsw = Array.from(g.normal(100))
  const sw = A.shapiroWilk(xsw)
  add({
    id: 'shapiroWilk',
    group: 'Normality',
    label: 'Shapiro–Wilk (n=100)',
    python: 'scipy',
    inputs: { x: xsw },
    columna: { statistic: sw.statistic, pValue: sw.pValue },
  })
  const mwA = Array.from(g.normal(120))
  const mwB = Array.from(g.normal(120, 0.25))
  const mw = A.mannWhitney(mwA, mwB)
  add({
    id: 'mannWhitney',
    group: 'Nonparametrics',
    label: 'Mann–Whitney U',
    python: 'scipy',
    inputs: { a: mwA, b: mwB },
    columna: { statistic: mw.statistic, pValue: mw.pValue },
  })
  const wx = Array.from(g.normal(80, 0.2))
  const wil = A.wilcoxonSigned(wx)
  const nNonZero = wx.filter((v) => v !== 0).length
  const wMinus = (nNonZero * (nNonZero + 1)) / 2 - wil.statistic
  add({
    id: 'wilcoxonSigned',
    group: 'Nonparametrics',
    label: 'Wilcoxon signed-rank',
    python: 'scipy',
    inputs: { x: wx },
    // Compare min(W+, W−) like scipy; p-value separately
    columna: { statistic: Math.min(wil.statistic, wMinus), pValue: wil.pValue },
    notes: 'columna exposes W+; comparison uses min(W+,W−) to match scipy',
  })
  const kg: Record<string, number[]> = {}
  for (let i = 0; i < 4; i++) kg[`g${i}`] = Array.from(g.normal(60, i * 0.3))
  const kr = A.kruskal(kg)
  add({
    id: 'kruskal',
    group: 'Nonparametrics',
    label: 'Kruskal–Wallis',
    python: 'scipy',
    inputs: { groups: kg },
    columna: { statistic: kr.statistic, pValue: kr.pValue },
  })
}

// ---- Regression ------------------------------------------------------------------------------------
{
  const n = 200
  const x1 = Array.from(g.normal(n))
  const x2 = Array.from(g.normal(n))
  const y = x1.map((v, i) => 1.5 + 2 * v - 0.7 * x2[i]! + 0.5 * g.normal(1)[0]!)
  const fit = A.ols(y, [x1, x2])
  add({
    id: 'ols',
    group: 'Regression',
    label: 'OLS + intercept (n=200, p=2)',
    python: 'numpy',
    inputs: { y, X: x1.map((v, i) => [v, x2[i]!]) },
    columna: {
      coefficients: fit.coefficients.map((c) => c.coef),
      r2: fit.r2,
      s: fit.s,
    },
  })
  const xg = Array.from(g.normal(150))
  const yg = xg.map((v) => (g.next() < 1 / (1 + Math.exp(-(0.2 + 1.2 * v))) ? 1 : 0))
  const gl = A.glm(yg, [xg], { family: 'binomial' })
  add({
    id: 'glm.binomial',
    group: 'Regression',
    label: 'GLM logit (n=150)',
    python: 'scipy',
    inputs: { y: yg, X: xg.map((v) => [v]) },
    columna: { coefficients: gl.coefficients.map((c) => c.coef) },
    notes: 'Python: scipy.optimize BFGS on Bernoulli log-likelihood',
  })
}

// ---- Time series / PCA / KM / linalg ---------------------------------------------------------------
{
  const ar: number[] = []
  let prev = 0
  for (let i = 0; i < 120; i++) {
    prev = 0.7 * prev + g.normal(1)[0]!
    ar.push(prev)
  }
  const ac = A.acf(ar, { maxLag: 5 })
  const arima = A.arima(ar, { p: 1, d: 0, q: 0, method: 'CSS-ML' })
  add({
    id: 'acf',
    group: 'Time series',
    label: 'ACF lags 0…5',
    python: 'numpy',
    inputs: { y: ar },
    columna: { acf: ac.acf.slice(0, 6) },
  })
  add({
    id: 'arima.ar1',
    group: 'Time series',
    label: 'ARIMA(1,0,0) φ',
    python: 'numpy',
    inputs: { y: ar },
    columna: { ar: arima.ar, intercept: arima.intercept },
    notes: 'numpy reference = OLS y_t ~ y_{t-1}; columna uses CSS-ML',
  })

  const z1 = Array.from(g.normal(80))
  const z2 = z1.map((v) => 0.8 * v + 0.6 * g.normal(1)[0]!)
  const z3 = Array.from(g.normal(80))
  const mat = z1.map((_, i) => [z1[i]!, z2[i]!, z3[i]!])
  const pc = A.pca(mat, { scale: true })
  add({
    id: 'pca',
    group: 'Multivariate',
    label: 'PCA eigenvalues (correlation)',
    python: 'numpy',
    inputs: { X: mat },
    columna: { eigenvalues: pc.eigenvalues },
  })

  const timeRaw = Array.from(g.exponential(60, 40))
  const censor = timeRaw.map((t) => (t > 50 ? 1 : 0))
  const time = timeRaw.map((t) => Math.min(t, 50))
  const km = A.kaplanMeier(time, { censor })
  add({
    id: 'kaplanMeier',
    group: 'Reliability',
    label: 'Kaplan–Meier S(t)',
    python: 'numpy',
    inputs: { time, censor },
    columna: {
      time: km.curve.map((p) => p.time),
      survival: km.curve.map((p) => p.survival),
    },
    notes: 'numpy: product-limit formula (no lifelines)',
  })

  const Amat = [
    [4, 1, 0.5],
    [1, 3, 0.2],
    [0.5, 0.2, 2],
  ]
  const bv = [1, 2, 0.5]
  const sol = A.lstsq(A.matrix(3, 3, Float64Array.from(Amat.flat())), bv)
  add({
    id: 'lstsq',
    group: 'Linear algebra',
    label: '3×3 lstsq',
    python: 'numpy',
    inputs: { A: Amat, b: bv },
    columna: { coef: Array.from(sol.coef) },
  })
}

mkdirSync(new URL('.', OUT), { recursive: true })
writeFileSync(
  OUT,
  JSON.stringify({
    generatedAt: new Date().toISOString(),
    seed: 20260915,
    nCases: cases.length,
    cases,
  }),
)
console.log(`Wrote ${cases.length} cases → ${OUT.pathname}`)

File: packages/bench/python/advanced_convergence.py
"""Compare columna/advanced numerical outputs to scipy / numpy on identical inputs.

Reads results/advanced-convergence-columna.json (from advanced-convergence.ts)
and writes docs/advanced-convergence.md.

    py -3 packages/bench/python/advanced_convergence.py
"""
from __future__ import annotations

import io
import json
import math
import sys
from pathlib import Path

import numpy as np
import scipy.stats as st
from scipy import optimize

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding="utf-8", line_buffering=True)

HERE = Path(__file__).resolve().parent
ROOT = HERE.parent.parent.parent
INP = HERE.parent / "results" / "advanced-convergence-columna.json"
OUT = ROOT / "docs" / "advanced-convergence.md"

AGREE = 1e-10
CLOSE = 1e-6
SOFT = 1e-3


def flatten(v):
    if isinstance(v, (list, tuple)):
        out = []
        for x in v:
            out.extend(flatten(x) if isinstance(x, (list, tuple)) else [float(x)])
        return out
    return [float(v)]


def max_abs_rel(a, b):
    aa = np.asarray(flatten(a), dtype=float)
    bb = np.asarray(flatten(b), dtype=float)
    n = min(aa.size, bb.size)
    if n == 0:
        return float("nan"), float("nan")
    aa, bb = aa[:n], bb[:n]
    # Align signs for statistics that may flip (t, z, U conventions)
    diff = np.abs(aa - bb)
    # Also try absolute values for signed test stats when relative gap is large
    diff_abs = np.abs(np.abs(aa) - np.abs(bb))
    if np.nanmax(diff_abs) < np.nanmax(diff) * 0.5:
        diff = diff_abs
        aa, bb = np.abs(aa), np.abs(bb)
    abs_err = float(np.nanmax(diff))
    rel = diff / np.maximum(1.0, np.abs(bb))
    rel_err = float(np.nanmax(rel))
    return abs_err, rel_err


def verdict(abs_err, rel_err):
    m = min(abs_err, rel_err) if math.isfinite(abs_err) and math.isfinite(rel_err) else float("inf")
    # Prefer relative for scale-free; for tiny refs use abs
    score = rel_err if abs_err > 1e-12 else abs_err
    if score <= AGREE or abs_err <= AGREE:
        return "agree"
    if score <= CLOSE or abs_err <= CLOSE:
        return "close"
    if score <= SOFT or abs_err <= SOFT:
        return "soft"
    return "diverge"


def km_numpy(time, censor):
    time = np.asarray(time, float)
    censor = np.asarray(censor, float)
    order = np.argsort(time)
    time, censor = time[order], censor[order]
    uniq = np.unique(time)
    S = 1.0
    times, surv = [], []
    for t in uniq:
        at = time >= t
        n_risk = int(at.sum())
        d = int(((time == t) & (censor == 0)).sum())
        if n_risk > 0 and d > 0:
            S *= 1.0 - d / n_risk
            times.append(float(t))
            surv.append(float(S))
    return times, surv


def acf_numpy(y, max_lag=5):
    y = np.asarray(y, float)
    y = y - y.mean()
    c0 = np.dot(y, y)
    out = [1.0]
    for k in range(1, max_lag + 1):
        out.append(float(np.dot(y[k:], y[:-k]) / c0))
    return out


def python_ref(case):
    cid = case["id"]
    inp = case["inputs"]
    if cid == "dist.normal.cdf":
        return {"values": st.norm.cdf(inp["x"]).tolist()}
    if cid == "dist.normal.ppf":
        return {"values": st.norm.ppf(inp["u"]).tolist()}
    if cid == "dist.t.cdf":
        return {"values": st.t.cdf(inp["x"], 10).tolist()}
    if cid == "dist.t.ppf":
        return {"values": st.t.ppf(inp["u"], 10).tolist()}
    if cid == "dist.chi2.sf":
        return {"values": st.chi2.sf(inp["x"], 5).tolist()}
    if cid == "dist.gamma.cdf":
        return {"values": st.gamma.cdf(inp["x"], 2.5, scale=3).tolist()}
    if cid == "dist.beta.ppf":
        return {"values": st.beta.ppf(inp["u"], 2, 5).tolist()}
    if cid == "dist.weibull.ppf":
        return {"values": st.weibull_min.ppf(inp["u"], 1.8, scale=50).tolist()}
    if cid == "dist.ptukey":
        return {"values": st.studentized_range.cdf(inp["q"], 4, 20).tolist()}
    if cid == "dist.qtukey":
        return {"values": st.studentized_range.ppf(inp["p"], 4, 20).tolist()}

    if cid == "ttest1":
        r = st.ttest_1samp(inp["x"], inp["mu"])
        return {"statistic": float(r.statistic), "pValue": float(r.pvalue), "estimate": float(np.mean(inp["x"]))}
    if cid == "ttest2":
        r = st.ttest_ind(inp["a"], inp["b"], equal_var=False)
        return {
            "statistic": float(r.statistic),
            "pValue": float(r.pvalue),
            "estimate": float(np.mean(inp["a"]) - np.mean(inp["b"])),
        }
    if cid == "ztest1":
        x = np.asarray(inp["x"], float)
        n = len(x)
        z = (x.mean() - inp["mu"]) / (inp["sigma"] / np.sqrt(n))
        return {"statistic": float(z), "pValue": float(2 * st.norm.sf(abs(z)))}
    if cid == "propTest1":
        r = st.binomtest(inp["k"], inp["n"], inp["p0"])
        return {"pValue": float(r.pvalue), "estimate": inp["k"] / inp["n"]}
    if cid == "propTest2.fisher":
        _, p = st.fisher_exact(inp["table"])
        return {"pValue": float(p)}
    if cid == "corrTest.pearson":
        r, p = st.pearsonr(inp["a"], inp["b"])
        # t statistic
        n = len(inp["a"])
        t = r * np.sqrt((n - 2) / max(1e-15, 1 - r * r))
        return {"estimate": float(r), "statistic": float(t), "pValue": float(p)}
    if cid == "corrTest.spearman":
        r, p = st.spearmanr(inp["a"], inp["b"])
        return {"estimate": float(r), "pValue": float(p)}
    if cid == "varTest1":
        x = np.asarray(inp["x"], float)
        n = len(x)
        s2 = x.var(ddof=1)
        chi = (n - 1) * s2 / (inp["sigma0"] ** 2)
        # two-sided like columna default
        p = 2 * min(st.chi2.cdf(chi, n - 1), st.chi2.sf(chi, n - 1))
        return {"statistic": float(chi), "pValue": float(p)}

    if cid in ("anova", "levene", "bartlett", "kruskal"):
        gs = [np.asarray(v, float) for v in inp["groups"].values()]
        if cid == "anova":
            r = st.f_oneway(*gs)
        elif cid == "levene":
            r = st.levene(*gs)
        elif cid == "bartlett":
            r = st.bartlett(*gs)
        else:
            r = st.kruskal(*gs)
        return {"statistic": float(r.statistic), "pValue": float(r.pvalue)}

    if cid == "chi2test":
        chi2, p, *_ = st.chi2_contingency(inp["table"], correction=False)
        return {"statistic": float(chi2), "pValue": float(p)}
    if cid == "chi2gof":
        o = np.asarray(inp["observed"], float)
        chi2, p = st.chisquare(o)
        return {"statistic": float(chi2), "pValue": float(p)}

    if cid == "andersonDarling":
        r = st.anderson(inp["x"], dist="norm")
        return {"statistic": float(r.statistic)}
    if cid == "shapiroWilk":
        w, p = st.shapiro(inp["x"])
        return {"statistic": float(w), "pValue": float(p)}
    if cid == "mannWhitney":
        r = st.mannwhitneyu(inp["a"], inp["b"], alternative="two-sided", method="asymptotic")
        return {"statistic": float(r.statistic), "pValue": float(r.pvalue)}
    if cid == "wilcoxonSigned":
        x = np.asarray(inp["x"], float)
        # columna reports W+; scipy.wilcoxon returns min(W+, W−) by default
        d = x[x != 0]
        ranks = st.rankdata(np.abs(d))
        w_plus = float(ranks[d > 0].sum())
        w_minus = float(ranks[d < 0].sum())
        r = st.wilcoxon(x, alternative="two-sided", method="approx")
        return {
            "statistic": float(min(w_plus, w_minus)),  # match scipy convention for comparison
            "pValue": float(r.pvalue),
            "_w_plus": w_plus,
        }

    if cid == "ols":
        y = np.asarray(inp["y"], float)
        X = np.asarray(inp["X"], float)
        Xd = np.column_stack([np.ones(len(y)), X])
        coef, *_ = np.linalg.lstsq(Xd, y, rcond=None)
        yhat = Xd @ coef
        ss_res = np.sum((y - yhat) ** 2)
        ss_tot = np.sum((y - y.mean()) ** 2)
        r2 = 1 - ss_res / ss_tot
        s = np.sqrt(ss_res / (len(y) - Xd.shape[1]))
        return {"coefficients": coef.tolist(), "r2": float(r2), "s": float(s)}

    if cid == "glm.binomial":
        y = np.asarray(inp["y"], float)
        X = np.asarray(inp["X"], float)
        Xd = np.column_stack([np.ones(len(y)), X])

        def nll(b):
            eta = Xd @ b
            eta = np.clip(eta, -30, 30)
            p = 1 / (1 + np.exp(-eta))
            return -np.sum(y * np.log(p + 1e-15) + (1 - y) * np.log(1 - p + 1e-15))

        b0 = np.zeros(Xd.shape[1])
        res = optimize.minimize(nll, b0, method="BFGS")
        return {"coefficients": res.x.tolist()}

    if cid == "acf":
        return {"acf": acf_numpy(inp["y"], 5)}
    if cid == "arima.ar1":
        y = np.asarray(inp["y"], float)
        # OLS with intercept: y_t = c + φ y_{t-1}
        yt = y[1:]
        yl = y[:-1]
        Xd = np.column_stack([np.ones(len(yt)), yl])
        coef, *_ = np.linalg.lstsq(Xd, yt, rcond=None)
        return {"ar": [float(coef[1])], "intercept": float(coef[0])}

    if cid == "pca":
        X = np.asarray(inp["X"], float)
        Z = (X - X.mean(0)) / X.std(0, ddof=1)
        # eigenvalues of correlation matrix / or via SVD of Z
        _, s, _ = np.linalg.svd(Z, full_matrices=False)
        eig = (s**2) / (X.shape[0] - 1)
        return {"eigenvalues": eig.tolist()}

    if cid == "kaplanMeier":
        t, s = km_numpy(inp["time"], inp["censor"])
        return {"time": t, "survival": s}

    if cid == "lstsq":
        coef, *_ = np.linalg.lstsq(np.asarray(inp["A"], float), np.asarray(inp["b"], float), rcond=None)
        return {"coef": coef.tolist()}

    raise KeyError(cid)


def main():
    data = json.loads(INP.read_text(encoding="utf-8"))
    rows = []
    counts = {"agree": 0, "close": 0, "soft": 0, "diverge": 0, "error": 0}

    for case in data["cases"]:
        try:
            ref = python_ref(case)
            col = case["columna"]
            # Compare overlapping keys
            abs_errs, rel_errs, details = [], [], []
            for k in col:
                if k not in ref or k.startswith("_"):
                    continue
                a_err, r_err = max_abs_rel(col[k], ref[k])
                abs_errs.append(a_err)
                rel_errs.append(r_err)
                details.append(f"{k}: abs={a_err:.2e} rel={r_err:.2e}")
            # Wilcoxon: also accept W+ vs scipy when only p matches
            if case["id"] == "wilcoxonSigned" and "_w_plus" in ref:
                a2, r2 = max_abs_rel(case["columna"].get("statistic"), ref["_w_plus"])
                if a2 < (abs_errs[0] if abs_errs else 1e9):
                    # already comparing min form from TS
                    pass
            if not abs_errs:
                raise RuntimeError("no overlapping keys")
            abs_err = max(abs_errs)
            rel_err = max(rel_errs)
            v = verdict(abs_err, rel_err)
            counts[v] += 1
            rows.append(
                {
                    "id": case["id"],
                    "group": case["group"],
                    "label": case["label"],
                    "python": case["python"],
                    "verdict": v,
                    "abs": abs_err,
                    "rel": rel_err,
                    "details": "; ".join(details),
                    "notes": case.get("notes") or "",
                }
            )
            print(f"{case['id']:28s} {v:8s} abs={abs_err:.3e} rel={rel_err:.3e}")
        except Exception as e:  # noqa: BLE001
            counts["error"] += 1
            rows.append(
                {
                    "id": case["id"],
                    "group": case["group"],
                    "label": case["label"],
                    "python": case["python"],
                    "verdict": "error",
                    "abs": float("nan"),
                    "rel": float("nan"),
                    "details": str(e),
                    "notes": case.get("notes") or "",
                }
            )
            print(f"{case['id']:28s} ERROR {e}")

    # Markdown report
    lines = []
    lines.append("# Numerical convergence: `columna/advanced` vs SciPy / NumPy")
    lines.append("")
    lines.append(
        f"Generated {data['generatedAt'][:10]} · seed {data['seed']} · "
        f"numpy {np.__version__}, scipy {getattr(__import__('scipy'), '__version__', '?')}."
    )
    lines.append("")
    lines.append(
        "Companion to [`advanced-benchmarks.md`](advanced-benchmarks.md) (throughput). "
        "Here the **same input arrays** are evaluated in columna and in SciPy/NumPy; "
        "we report max absolute and relative differences on the compared scalars/vectors."
    )
    lines.append("")
    lines.append("### Verdict thresholds")
    lines.append("")
    lines.append("| Verdict | Criterion (max abs **or** max rel error) |")
    lines.append("|---|---|")
    lines.append(f"| **agree** | ≤ {AGREE:g} |")
    lines.append(f"| **close** | ≤ {CLOSE:g} |")
    lines.append(f"| **soft** | ≤ {SOFT:g} |")
    lines.append("| **diverge** | above soft |")
    lines.append("")
    lines.append(
        "Signed test statistics are compared on absolute values when that reduces the gap "
        "(Welch t / Mann–Whitney U sign conventions)."
    )
    lines.append("")
    lines.append("Regenerate: `pnpm --filter @columna/bench run advanced:convergence`.")
    lines.append("")

    # Summary
    n = len(rows)
    lines.append("## Summary")
    lines.append("")
    lines.append(
        f"- Cases: **{n}** · agree **{counts['agree']}** · close **{counts['close']}** · "
        f"soft **{counts['soft']}** · diverge **{counts['diverge']}** · error **{counts['error']}**"
    )
    good = counts["agree"] + counts["close"]
    lines.append(
        f"- Within close (≤1e-6): **{good}/{n}** ({100 * good / max(1, n):.0f}%)."
    )
    lines.append("")

    groups = []
    for r in rows:
        if not groups or groups[-1] != r["group"]:
            groups.append(r["group"])

    for group in groups:
        lines.append(f"## {group}")
        lines.append("")
        lines.append("| Case | Library | Verdict | max ‖Δ‖ | max rel | Notes |")
        lines.append("|---|---|---|---:|---:|---|")
        for r in rows:
            if r["group"] != group:
                continue
            abs_s = "—" if not math.isfinite(r["abs"]) else f"{r['abs']:.2e}"
            rel_s = "—" if not math.isfinite(r["rel"]) else f"{r['rel']:.2e}"
            note = r["notes"] or r["details"]
            if len(note) > 80:
                note = note[:77] + "…"
            lines.append(
                f"| `{r['id']}` — {r['label']} | {r['python']} | **{r['verdict']}** | {abs_s} | {rel_s} | {note} |"
            )
        lines.append("")

    # Divergences detail
    bad = [r for r in rows if r["verdict"] in ("diverge", "soft", "error")]
    if bad:
        lines.append("## Soft / diverge / error detail")
        lines.append("")
        for r in bad:
            lines.append(f"### `{r['id']}` ({r['verdict']})")
            lines.append("")
            lines.append(f"- {r['details']}")
            if r["notes"]:
                lines.append(f"- Note: {r['notes']}")
            lines.append("")

    lines.append("## Method notes")
    lines.append("")
    lines.append(
        "- Distributions: SciPy `scipy.stats` CDF/PPF/SF on the same grid as columna."
    )
    lines.append(
        "- Hypothesis tests: SciPy defaults closest to columna (Welch `ttest_ind`, "
        "`binomtest`, `fisher_exact`, asymptotic Mann–Whitney)."
    )
    lines.append("- OLS / ACF / AR(1) / PCA / KM / lstsq: NumPy reference formulas.")
    lines.append(
        "- GLM logit: SciPy BFGS on Bernoulli NLL (not statsmodels); expect soft agreement."
    )
    lines.append(
        "- ARIMA CSS-ML vs OLS AR(1): intentional estimator difference — soft/diverge is informative, not a bug by itself."
    )
    lines.append("")

    OUT.write_text("\n".join(lines), encoding="utf-8")
    print(f"\nWrote {OUT}")


if __name__ == "__main__":
    main()
